Context, Params and Variables guides (#2442)
astronomer/docs: learn/airflow-variables.md

---
title: "Use Airflow Variables"
sidebar_label: "Variables"
description: "Create and use Airflow Variables."
id: airflow-variables
---

> An Airflow variable is a key-value pair that can be used to store information in your Airflow environment. They are commonly used to store instance level information that rarely changes, including secrets like for example an API key or the path to a configuration file. Airflow variables are [Fernet](https://github.com/fernet/spec/) encrypted and can also be retrieved from a [secrets backend](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/security/secrets/secrets-backend/index.html).

Reviewer: Somewhere in here we should link to our Astro variables docs, maybe in an info box. I think this is most relevant: https://docs.astronomer.io/astro/environment-variables

There are two distinct types of Airflow variables, regular values and JSON serialized values. 

This concept guide covers how to create Airflow variables and access them programmatically.

## Assumed knowledge

To get the most out of this guide, you should have an understanding of:

- Airflow DAGs. See [Introduction to Airflow DAGs](dags.md).
- Airflow operators. See [Operators 101](what-is-an-operator.md).

## Best practices for storing information in Airflow

Airflow offers several ways to store your variables. The ideal option often depends on what type of information you are storing and where and how you want to access it:

> - Environment variables are a common way to store small pieces of information to be available in your whole Airflow environment. The advantage of using environment variables is that you can include their creation in your CI/CD process. Environment variables are the only type of variable storage that can be accessed in top-level DAG code. Note that unless you create an Airflow variable via the environment variable by starting it with `AIRFLOW_VAR_` you will not be able to view or update it in the Airflow UI. 

Author: Yes haha, definitely the latter. Will try to clarify 😅

> - Airflow variables are ideal to store key-value pairs or short JSON objects that need to be accessible in your whole Airflow environment. It is best practice to use Airflow variables for information that is runtime dependent but does not change too frequently. Airflow variables are also encrypted making them suitable to be used for storing secrets.

Reviewer: ```suggestion
- Airflow variables store key-value pairs or short JSON objects that need to be accessible in your whole Airflow environment. It is best practice to use Airflow variables for information that is runtime dependent but does not change too frequently. Airflow variables are also encrypted, making them suitable for storing secrets.
```

![UI](/img/guides/airflow-variables_UI.png)

The Airflow CLI contains options to set, get and delete [Airflow variables](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#variables). To create a variable via the CLI use the following command:

<Tabs
    defaultValue="astro"
    groupId= "create-airflow-variable"
    values={[
        {label: 'Astro CLI', value: 'astro'},
        {label: 'Airflow CLI', value: 'airflow'},
    ]}>
<TabItem value="astro">

```sh
astro dev run variables set my_var my_value
astro dev run variables set -j my_json_var '{"key": "value"}'
```

</TabItem>

<TabItem value="airflow">

```sh
airflow variables set my_var my_value
airflow variables set -j my_json_var '{"key": "value"}'
```

</TabItem>

</Tabs>

To set Airflow variables via an environment variable create an environment variable with the prefix `AIRFLOW_VAR_` + the name of the variable you want to set

```text
AIRFLOW_VAR_MYREGULARVAR='my_value'
AIRFLOW_VAR_MYJSONVAR='{"hello":"world"}'
```

> Astro CLI users can create Airflow variables from the `airflow_settings.yaml` file. 

Reviewer: Link to CLI docs on "Astro CLI" here

```yaml
  variables:
    - variable_name: my_regular_variable_key
      variable_value: my_value
    - variable_name: my_json_variable_key
      variable_value: '{"hello":"world"}'
```

Lastly, you can programmatically set Variables within your Airflow tasks via the [Variable model](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/variable/index.html#module-airflow.models.variable). If you want to serialize a JSON value, make sure to set `serialize_json=True`.

```python
@task
def set_var():
    from airflow.models import Variable
    Variable.set(key="my_regular_var", value="Hello!")
    Variable.set(key="my_json_var", value={"num1": 23, "num2": 42}, serialize_json=True)
```

Updating an Airflow variable works the same way by using the `.update()` method.

## Retrieving an Airflow variable

> To programmatically retrieve and Airflow variable you can either use the `.get()` method of the Airflow Variable model or you can pull the Variable value directly from the [Airflow context](airflow-context).

Reviewer: ```suggestion
To programmatically retrieve an Airflow variable, you can either use the `.get()` method of the Airflow Variable model or you can pull the Variable value directly from the [Airflow context](airflow-context).
```

---
title: "Use Airflow Variables"
sidebar_label: "Variables"
description: "Create and use Airflow Variables."
id: airflow-variables
---

An Airflow variable is a key-value pair that can be used to store information in your Airflow environment. They are commonly used to store instance level information that rarely changes, including secrets like for example an API key or the path to a configuration file. Airflow variables are [Fernet](https://github.com/fernet/spec/) encrypted and can also be retrieved from a [secrets backend](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/security/secrets/secrets-backend/index.html).

There are two distinct types of Airflow variables, regular values and JSON serialized values. 

This concept guide covers how to create Airflow variables and access them programmatically.

## Assumed knowledge

To get the most out of this guide, you should have an understanding of:

- Airflow DAGs. See [Introduction to Airflow DAGs](dags.md).
- Airflow operators. See [Operators 101](what-is-an-operator.md).

> ## Best practices for storing information in Airflow

Reviewer: Since this guide is about Airflow variables, and there's often a lot of confusion about Airflow vs environment variables, I'm thinking we should reorg this section a bit to start with the text about Airflow variables, and then have something like "There are also other ways of storing information in Airflow, including:" and then the rest of the bullets. And make sure we're super clear in the env var bullet what the difference is.

:::info

To learn more about how to set environment variables on Astro, see [Environment Variables](https://docs.astronomer.io/astro/environment-variables).

:::

## Create an Airflow variable

There are several ways to create Airflow variables:

- Using the Airflow UI
- Using the Airflow CLI.
- Using an environment variable.
- Using the `airflow_settings.yaml` file (Astro CLI users).
- Programmatically from within an Airflow task.

To create an Airflow Variable in the UI, click on the **Admin** tab and select **Variables**. Then click on the **+** button and enter a key, value and an optional description for your variable. You also have the option to **Import Variables** from a file.

![UI](/img/guides/airflow-variables_UI.png)

The Airflow CLI contains options to set, get and delete [Airflow variables](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#variables). To create a variable via the CLI use the following command:

> <Tabs

Reviewer: I'm not seeing the tabs render for these in the deploy preview...

- Airflow Variables should be used for information that is runtime dependent but doesn't change too frequently.
- You should avoid using Airflow Variables outside of tasks in top-level DAG code, as they will create a connection to the Airflow metastore every time the DAG is parsed, which can lead to performance issues. See [DAG writing best practices in Apache Airflow](dag-best-practices.md#avoid-top-level-code-in-your-dag-file).
- If you do use Airflow Variables in top-level DAG code, use the [Jinja template](templating.md) syntax so that your Variables are only rendered when a task executes.
- Airflow Variables are encrypted with [Fernet](https://github.com/fernet/spec/) when they are written to the Airflow metastore. To mask Airflow Variables in the UI and logs, include a substring indicating a sensitive value in your Airflow Variable name. See [Hiding sensitive information](#hiding-sensitive-information).

See the Airflow documentation for examples of code showing [good and bad practices for accessing Airflow Variables in a DAG](https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#airflow-variables).

> Aside from Airflow Variables, there are other ways of storing information in Airflow. The ideal option often depends on what type of information you are storing and where and how you want to access it:

Reviewer 2: ```suggestion
Aside from Airflow Variables, there are other ways of storing information in Airflow. The ideal option depends on what type of information you are storing and where and how you want to access it:
```

There are two distinct types of Airflow variables, regular values and JSON serialized values. 

![Variables in the Airflow UI](/img/guides/airflow-variables_variables_in_UI.png)

This concept guide covers how to create Airflow variables and access them programmatically.

## Assumed knowledge

To get the most out of this guide, you should have an understanding of:

- Airflow DAGs. See [Introduction to Airflow DAGs](dags.md).
- Airflow operators. See [Operators 101](what-is-an-operator.md).

## Best practices for storing information in Airflow

[Airflow variables](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/variables.html#variables) store key-value pairs or short JSON objects that need to be accessible in your whole Airflow instance. They are are Airflow’s runtime configuration concept and defined using the [`airflow.model.variable`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/variable/index.html#module-airflow.models.variable) object. 

There are some best practices to keep in mind when using Airflow variables:

- Airflow variables should be used for information that is runtime dependent but does not change too frequently.

> - You should avoid using Variables in top-level Airflow code, as they will create a connection to the Airflow metastore every time the DAG is parsed, which can lead to performance issues. See also [DAG writing best practices in Apache Airflow](dag-best-practices.md#avoid-top-level-code-in-your-dag-file).

Reviewer: ```suggestion
- You should avoid using Airflow variables in top-level DAG code, as they will create a connection to the Airflow metastore every time the DAG is parsed, which can lead to performance issues. See also [DAG writing best practices in Apache Airflow](dag-best-practices.md#avoid-top-level-code-in-your-dag-file).
```

> - You should avoid using Airflow variables in top-level DAG code, as they will create a connection to the Airflow metastore every time the DAG is parsed, which can lead to performance issues. See also [DAG writing best practices in Apache Airflow](dag-best-practices.md#avoid-top-level-code-in-your-dag-file).

Reviewer 2: ```suggestion
- You should avoid using Airflow variables outside of tasks in top-level DAG code, as they will create a connection to the Airflow metastore every time the DAG is parsed, which can lead to performance issues. See [DAG writing best practices in Apache Airflow](dag-best-practices.md#avoid-top-level-code-in-your-dag-file).
```

> - You can use the [Jinja template](templating.md) syntax to access Variables in top-level Airflow code, since they will only be rendered one a task executes.

Reviewer: ```suggestion
- You can use the [Jinja template](templating.md) syntax to access Airflow variables in top-level DAG code, since they will only be rendered once a task executes.
```

- [Params](airflow-params.md) can be used to store variables that are specific to a DAG or DAG run. You can define defaults for params at the DAG or task level and override them at runtime. Params are not encrypted and should not be used to store secrets.
- [XComs](airflow-passing-data-between-tasks.md) can be used to pass small pieces of information between Airflow tasks. Use XComs when the information is likely to change with each DAG run and mostly needs to be accessed by individual tasks in or outside of the DAG from within which the XCom is created. Default XComs are not encrypted and should not be used to store secrets.

> :::info
> 
> To learn more about how to set environment variables on Astro, see [Environment Variables](https://docs.astronomer.io/astro/environment-variables).
> 
> :::

Reviewer 2: ```suggestion
```

Feels out of place, I would drop this link later when talking about how you can set Airflow variables.

## Create an Airflow variable

There are several ways to create Airflow variables:

- Using the Airflow UI
- Using the Airflow CLI.
- Using an environment variable.

> - Using the `airflow_settings.yaml` file (Astro CLI users).

Reviewer 2: ```suggestion
```
We decided not to mention this option in our own Astro variables and connection management doc because there just isn't a great user case for it right now. So I would also remove it from this guide.

```text
AIRFLOW_VAR_MYREGULARVAR='my_value'
AIRFLOW_VAR_MYJSONVAR='{"hello":"world"}'
```

[Astro CLI](https://docs.astronomer.io/astro/cli/install-cli) users can create Airflow variables from the `airflow_settings.yaml` file. 

```yaml
  variables:
    - variable_name: my_regular_variable_key
      variable_value: my_value
    - variable_name: my_json_variable_key
      variable_value: '{"hello":"world"}'
```

Lastly, you can programmatically set Variables within your Airflow tasks via the [Variable model](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/variable/index.html#module-airflow.models.variable). If you want to serialize a JSON value, make sure to set `serialize_json=True`.

```python
@task
def set_var():
    from airflow.models import Variable
    Variable.set(key="my_regular_var", value="Hello!")
    Variable.set(key="my_json_var", value={"num1": 23, "num2": 42}, serialize_json=True)
```

Updating an Airflow variable works the same way by using the `.update()` method.

## Retrieving an Airflow variable

> To programmatically retrieve an Airflow variable, you can either use the `.get()` method of the Airflow Variable model or you can pull the Variable value directly from the [Airflow context](airflow-context).

Author: uppercased as per our Slack discussion :)

    )["num1"]
    print(my_regular_var)
    print(my_json_var)

@task
def get_var_from_context(**context):
    my_regular_var = context["var"]["value"].get("my_regular_var")
    my_json_var = context["var"]["json"].get("my_json_var")["num2"]
    print(my_regular_var)
    print(my_json_var)
```

For using Airflow variables in traditional Airflow operators it is often easier to use a [Jinja template](templating.md). See also [Airflow Variables in Templates](https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html#airflow-variables-in-templates).

```python
get_var_jinja = BashOperator(
    task_id="get_var_jinja",
    bash_command='echo "{{ var.value.my_regular_var }} {{ var.json.my_json_var.num2 }}"',
)
```

You can also retrieve an Airflow variable via the Airflow CLI's [`get`](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#get_repeat3) and [`list`](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#list_repeat8) commands. 

## Hiding sensitive information

Airflow variables are [Fernet](https://github.com/fernet/spec/) encrypted in the Airflow metastore.

> As seen in the screenshot at the beginning of this guide, some Airflow variables are additionally masked in the Airflow UI and if printed to the logs. This behavior is controlled via the Airflow configuration. By default, the [`hide_sensitive_var_conn_fields`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#hide-sensitive-var-conn-fields) configuration is set to `True`, which masks all Airflow variables that contain the following strings:

Reviewer 2: ```suggestion
As seen in the screenshot at the beginning of this guide, some Airflow variables are additionally masked in the Airflow UI and logs. While you can manually mark variables as secrets, some variables are marked automatically using an Airflow configuration. By default, the [`hide_sensitive_var_conn_fields`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#hide-sensitive-var-conn-fields) configuration is set to `True`, which automatically masks all Airflow variables that contain the following strings:
```

- `access_token`
- `api_key`
- `apikey`
- `authorization`
- `passphrase`
- `passwd`
- `password`
- `private_key`
- `secret`
- `token`

> This list can be extended by adding comma separated strings to the [`sensitive_var_conn_names`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#sensitive-var-conn-names) configuration.
> 
> See also [Masking sensitive data](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/security/secrets/mask-sensitive-values.html).

Reviewer 2: ```suggestion
This list can be extended by adding comma separated strings to the [`sensitive_var_conn_names`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#sensitive-var-conn-names) configuration. See [Masking sensitive data](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/security/secrets/mask-sensitive-values.html).
```
